deeplake/core/vectorstore/deep_memory.py

```diff
@@ -0,0 +1,457 @@
+import uuid
+from typing import Any, Dict, Optional, List, Union, Callable, Tuple
+from time import time
+
+import numpy as np
+
+try:
+    from indra import api
+
+    INDRA_INSTALLED = True
+except Exception:
+    INDRA_INSTALLED = False
+
+import deeplake
+from deeplake.enterprise.dataloader import indra_available
+from deeplake.constants import DEFAULT_QUERIES_VECTORSTORE_TENSORS
+from deeplake.core.dataset import Dataset
+from deeplake.core.vectorstore.deeplake_vectorstore import VectorStore
+from deeplake.client.client import DeepMemoryBackendClient
+
+
+class DeepMemory:
+    def __init__(
+        self,
+        dataset: Dataset,
+        embedding_function: Optional[Any] = None,
+        token: Optional[str] = None,
+    ):
+        """Based Deep Memory class to train and evaluate models on DeepMemory managed service.
+
+        Args:
+            dataset (Dataset): deeplake dataset object.
+            embedding_function (Optional[Any], optional): Embedding funtion class used to convert queries/documents to embeddings. Defaults to None.
+            token (Optional[str], optional): API token for the DeepMemory managed service. Defaults to None.
+
+        Raises:
+            ImportError: if indra is not installed
+        """
+        self.dataset = dataset
+        self.token = token
+        self.embedding_function = embedding_function
+        self.client = DeepMemoryBackendClient(token=token)
+        self.queries_dataset = deeplake.dataset(
+            self.dataset.path + "_eval_queries",
+            token=token,
+            read_only=False,
+        )
+        if len(self.queries_dataset) == 0:
+            self.queries_dataset.commit(allow_empty=True)
+
+    def train(
+        self,
+        queries: List[str],
+        relevance: List[List[Tuple[str, int]]],
+        embedding_function: Optional[Callable[[str], np.ndarray]] = None,
+        token: Optional[str] = None,
+    ) -> str:
+        """Train a model on DeepMemory managed service.
+
+        Examples:
+            >>> queries: List[str] = ["What is the capital of India?", "What is the capital of France?"]
+            >>> relevance: List[List[Tuple[str, int]]] = [[("doc_1", 1), ("doc_2", 1)], [("doc_3", 1)]]
+            >>> job_id: str = vectorstore.deep_memory.train(queries, relevance)
+
+        Args:
+            queries (List[str]): List of queries to train the model on.
+            relevance (List[List[Tuple[str, int]]]): List of relevant documents for each query with their respective relevance score.
```

Comment:
@adolkhan I'm a bit confused by the Examples and the definition of the relevance parameter. Presumably each document in this parameter is connected to the corpus via some id? If so, that needs to be super clear. Are "doc_1", "doc_2", .... etc ids? Do they map to a tensor in the corpus. If so, which tensor?

IMO this description/examples are vague and non-generalizable. Pls make them more specific, so it's crystal clear how the relevante information is linked to the corpus.

Reply:
Is it clearer now?

Reply:
@adolkhan I don't see any difference. You sure you pushed something?